### PR TITLE
Smart Contract Interpreter versionning

### PR DESCRIPTION
**Pull request by bchamagne**
# Description

Introduce smart contract versioning. 
The user will be able to choose it's version by setting a `@version` attribute at the top of her/his code. No `@version` attributes is equivalent to 0.0.1.
Part of #894

There are 2 versions available:
1. legacy / functional lang (= 0.0.1)
2. imperative lang (= 1.0.0)

I move current language into a version0 folder. 
The imperative lang is not implemented in this PR but I prepared the ground in a version1 interpreter. 
Some files has been renamed to match their modules names.

## TODO
- [x] Unplug the worker from version0 and plug it to the dispatcher (Archethic.Contracts.Interpreter) instead. 
- [x] Unplug the contracts from version0 and plug it to the dispatcher instead.
- [x] Should I restrict versions to only those existing? 

## Examples

```
@version "1.0.0" 
condition ....
actions ...
```

```
@version "0.0.1"
condition ...
actions ...
```

```
condition ...
actions ...
```

# Type of change

- New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Unit tests :heavy_check_mark: 
Legacy smart contract  :heavy_check_mark: 

# Checklist:

- My code follows the style guidelines of this project
- I have performed a self-review of my own code
- I have commented my code, particularly in hard-to-understand areas
- I have made corresponding changes to the documentation
- My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature works
- New and existing unit tests pass locally with my changes
- Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by samuelmanzanera** [comment]
Is this PR still relevant regarding #894

**Comment by bchamagne** [comment]
> Is this PR still relevant regarding #894 ?

yes, this is step 1/2 (step 2 is https://github.com/archethic-foundation/archethic-node/pull/903). 
This one prepared the ground to have 2 different interpreters

**Comment by samuelmanzanera** [comment]
I wonder about the granularity of the versions.
The last digit might not be useful, as the bug fixes will be patched along the way, and the contract code might not be dependent on the bug fixes.
For the second one, I think the enhancements of the interpreter should be retrocompatible within a major version.
But the one is useful because it informs which major version of the interpreter is used and will simplify the writing of the smart contract code.

File: lib/archethic/contracts/interpreter.ex [hunk 1]
```diff
@@ -3,241 +3,155 @@ defmodule Archethic.Contracts.Interpreter do
...
-  defp parse_contract({:__block__, _, ast}, contract) do
-    parse_ast_block(ast, contract)
+  @doc """
+  Determine from the code, the version to use.
+  Return the version & the code where the version has been removed.
+  (should be private, but there are unit tests)
+  """
+  @spec version(String.t()) :: {version(), String.t()} | :error
+  def version(code) do
+    regex_opts = [capture: :all_but_first]
```
Comment (samuelmanzanera) [inline lib/archethic/contracts/interpreter.ex:52]: Why regex instead of the AST ?

Comment (bchamagne) [reply lib/archethic/contracts/interpreter.ex]: much cleaner, thanks!

**Comment by bchamagne** [comment]
reminder, once merged we need an upgrade to the playground (see: https://github.com/archethic-foundation/smart-contract-playground/pull/21/files#diff-2e3dab7734e486531bea720f3148534abd3b883ae941d433cdf70078a41b2643)

**Review by Neylix — APPROVED** [review]

**Pull request merged**